ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java

```diff
@@ -162,48 +176,61 @@ void handleResponse(ResponseInfo responseInfo) {
           logger.error("The correlation id in the DeleteResponse " + deleteResponse.getCorrelationId()
               + " is not the same as the correlation id in the associated DeleteRequest: " + deleteRequest
               .getCorrelationId());
+          // @todo Discuss why this would complete the operation.
+          routerMetrics.unknownReplicaResponseError.inc();
           setOperationException(
               new RouterException("Received wrong response that is not for the corresponding request.",
                   RouterErrorCode.UnexpectedInternalError));
-          updateOperationState(replica, RouterErrorCode.UnexpectedInternalError);
-          return;
+          routerErrorCode = RouterErrorCode.UnexpectedInternalError;
+          updateOperationState(replica, routerErrorCode);
         } else {
           responseHandler.onRequestResponseError(replica, deleteResponse.getError());
-          processServerError(replica, deleteResponse.getError());
+          // The status of operation tracker will be updated within the processServerError method.
+          routerErrorCode = processServerError(replica, deleteResponse.getError());
         }
       } catch (IOException e) {
-        // @todo: Even this should really not happen. But we need a metric.
         logger.error("Unable to recover a deleteResponse from received stream.");
-        updateOperationState(replica, RouterErrorCode.UnexpectedInternalError);
+        routerErrorCode = RouterErrorCode.UnexpectedInternalError;
+        updateOperationState(replica, routerErrorCode);
       }
     }
     checkAndMaybeComplete();
+    if (routerErrorCode != null) {
+      dataNodeBasedMetrics.deleteRequestErrorCount.inc();
+    }
   }
 
   /**
    * A wrapper class that is used to check if a request has been expired.
    */
-  private class InflightRequestInfo {
-    final long submissionTime;
+  private class DeleteRequestInfo {
+    final long startTimeMs;
```

Comment:
Have changed to directly access inner class member variables from the outer class (for both Delete and Put).